### ISSUE TITLE
do not wait making available configuration.set/get until json file is loaded

### ISSUE THREAD

**Issue by timomayer** [biotope/biotope-configuration#14]
current implementation waits until configuration.json file is loaded (async) until the configuration getter and setter classes can be used.

this makes it quite hard to work with the configuration because you have to always check if configuration.set or get is available or you have to delay execution of all component js code until configuration is ready.

i suggest to rewrite the configuration to provide setter and getter as soon as the configuration is initialized. load json async and add values from json to configuration when ready.

problem: what to do when we try to access a property from configuration.json but json is not yet loaded and value not yet set. solution
if configuration.get is called while a configuration.json file is still loading delay get response until json file is loaded.

**Comment by janrembold** [comment]
@timomayer What problem are you trying to solve here? In my eyes this creates as much problems as it solves. 

For me a configuration plugin is only a singleton object with getter and setter methods that should be loaded as soon as possible in the head section. Footprint should be as small as possible.

If there is any configuration file that holds global configuration values that should be available when document is ready, it should be loaded directly after the configuration plugin without any fancy async AJAX calls. 

It would be helpful to add a target and/or a list of problems that this plugin should solve in the readme section.

**Comment by timomayer** [comment]
i agree with you the only problem i see is performance.
we need the configuration getter / setter in the head section so we need to load the configuration class there - not an issue. but when we load the global configuration file there too and in sync that means the whole rendering of the DOM is delayed until the configuration.json is loaded and i think this is an issue for overall performance.

**Comment by smohadjer** [comment]
I agree with Jan on that we are losing track on what we are trying to achieve with this script and I think this is a typical case of writing code without a spec, but I also agree with Timo that his solution is what we need right now. Since we have no spec (I seriously think we should start requiring a spec/documentation to be added to every tool or library we write to describe the reason for writing it in the first place and what it should or should not do, and so on), here is a bit of history on this tool:

The original problem that brought us to write this script was that a client wanted ability to overwrite some settings in a site's default configuration file on certain pages by setting them in head of those pages. At that time settings in head were being overwritten by configuration file (a js file then, a json now) as that file was being feteched and run before closing body tag to avoid affecting performance. So we needed a script to fetch the configuration file and all the preexisting configuration settings (set in page head or anywhere else) and merge them into one configuration object. The first version of script which was just couple of lines of code resolved this problem. 

But then came a revison where set and get methods were added and that's when things started to get complicated. The idea was that after configuration script is done these methods should be used to interact with config settings and overwriting config settings directly should not be possible anymore. Until this point we didn't really care WHEN configuration settings are set. Whether they were set before or after script runs, they always updated properties in configuration file. Still we added these methods and had no problem as far they were only being used after configuration script was done. But now we have decided to use these methods even before configuration script is done, mainly I guess for sake of consistency, and that brings us to where we are now. 

So I think as Timo suggested cofiguration object should be initialized as the first thing on a page and during initialization only set and get methods are added. Then we should add a new method, e.g. loadJson() that is invoked after all external scripts are loaded to fetch the configuration.json and update it with preexisting settings. We also need to make sure in none of the projects using this tool configuration parameters are set directly as is the case in handlebars.template.js in one of the porjects right now.

**Comment by timomayer** [comment]
@saeid we have the rfcs now for keeping track on that (we haven't had that when we rewrote the configuration).

Let's do the following changes please:
- make possible to initiate config to have get and set without loading a file
- provide function to load config json file (with option for merge strategy)
Make json file load sync instead of async

**Comment by timomayer** [comment]
I agree async is better but then we have to make sure a configuration.get directly called after jsonLoad blocks until jsonLoad is ready and config values are set. I don't want to have the devs care about is the configuration ready or not

**Comment by smohadjer** [comment]
The issue is not only when the first get() is used, but also when the first set() is used. If any of these methods is called while json is loading we should wait until json is loaded and configuration data has been updated before procceding with code execution.

**Comment by timomayer** [comment]
yes you are right it is get and set. do you agree waiting when getter or setter is called is a litte bit more complex inside the configuration class but much easier in handling for developers? and a possible performance gain when first getter or setter is called a lot of lines later after jsonLoad and the lines in between are executed in the meantime?

**Comment by smohadjer** [comment]
Yes, I think this is a better approach.

**Comment by smohadjer** [comment]
@timomayer Waiting when the first getter/setter is called may not just be a bit more complex, but perhaps even impossible! I assume waiting at that point means we have to block code execution (via a while loop for example) and if we do so, then call stack will never get empty, so browser won't have a chance to process the next event in the queue and find out whether ajax/fetch is done or not. Or did you have a different approach in mind?

I honestly think stopping code execution regardless of when we do it is a bad idea. I suggest we simply initialize configurationLoader in head and before any other script runs so getter/setter can be used immediately. Our json will load async, but we don't initialize site/app script until json is fetched and configuration data is ready. This way developers don't need to care whether configuration is ready or not since their code never runs before it's ready and we never freeze the browser either. Can you think of a situation in which we may run into problem using such strategy?

**Comment by timomayer** [comment]
think it is possible to do that via https://javascript.info/async-await and a polyfill for older browsers? what do you think?

**Comment by smohadjer** [comment]
No, it's not. I have opened a pull request and updated version to 3.0.0 since this version is not backward compatible.
